<a id="table-of-content"></a>
# Table of Content
- [Introduction](#1)
    - About this Notebook
    - About this Dataset
- [Preparation](#2)
    - [Drop unused Columns](#2.1)
    - [Repair Columns in wrong type](#2.2)
- [Data Overview](#3)
    - [Quick Look](#3.1)
    - [Basic Statistics](#3.2)

<a id='1'></a>
# 1 Introduction
## 1.1 About this Notebook
- This Notebook is my DS pratice project on `10/09/2021`, mainly about EDA (Exploratory Data Analysis).

## 1.2 About this Dataset
The dataset include all the songs that have been on the **Top 200 Weekly (Global) charts of Spotify in 2020 & 2021**. The dataset include the following features:
   - `Highest Charting Position`: The highest position that the song has been on in the Spotify Top 200 Weekly Global Charts in 2020 & 2021.
   - `Number of Times Charted`: The number of times that the song has been on in the Spotify Top 200 Weekly Global Charts in 2020 & 2021.
   - `Week of Highest Charting`: The week when the song had the Highest Position in the Spotify Top 200 Weekly Global Charts in 2020 & 2021.
   - `Song Name`: Name of the song that has been on in the Spotify Top 200 Weekly Global Charts in 2020 & 2021.
   - `Song iD`: The song ID provided by Spotify (unique to each song).
   - `Streams`: Approximate number of streams the song has.
   - `Artist`: The main artist/ artists involved in making the song.
   - `Artist Followers`: The number of followers the main artist has on Spotify.
   - `Genre`: The genres the song belongs to.
   - `Release Date`: The initial date that the song was released.
   - `Weeks Charted`: The weeks that the song has been on in the Spotify Top 200 Weekly Global Charts in 2020 & 2021.
   - `Popularity`:The popularity of the track. The value will be between 0 and 100, with 100 being the most popular.
   - `Danceability`: Danceability describes how suitable a track is for dancing based on a combination of musical elements including tempo, rhythm stability, beat strength, and overall regularity. A value of 0.0 is least danceable and 1.0 is most danceable.
   - `Acousticness`: A measure from 0.0 to 1.0 of whether the track is acoustic.
   - `Energy`: Energy is a measure from 0.0 to 1.0 and represents a perceptual measure of intensity and activity. Typically, energetic tracks feel fast, loud, and noisy.
   - `Instrumentalness`: Predicts whether a track contains no vocals. The closer the instrumentalness value is to 1.0, the greater likelihood the track contains no vocal content.
   - `Liveness`: Detects the presence of an audience in the recording. Higher liveness values represent an increased probability that the track was performed live.
   - `Loudness`: The overall loudness of a track in decibels (dB). Loudness values are averaged across the entire track. Values typical range between -60 and 0 db.
   - `Speechiness`: Speechiness detects the presence of spoken words in a track. The more exclusively speech-like the recording (e.g. talk show, audio book, poetry), the closer to 1.0 the attribute value.
   - `Tempo`: The overall estimated tempo of a track in beats per minute (BPM). In musical terminology, tempo is the speed or pace of a given piece and derives directly from the average beat duration.
   - `Valence`: A measure from 0.0 to 1.0 describing the musical positiveness conveyed by a track. Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry).
   - `Chord`: The main chord of the song instrumental.

### Acknowledgements
This dataset would not be possible without the help of spotifycharts.com and Spotipy Python Library

<a id='2'></a>
# 2 Preparation
- Prepare the packages and data that will be used for the analysis process.
- This dataset contains 1 file.

In [1]:
# import packages
import os
import pandas as pd

import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

# Some pandas display options
pd.set_option('display.max_row', None)
pd.set_option('display.max_column', None)
pd.set_option('float_format', '{:.2f}'.format)

# Load the Dataset
df = pd.read_csv('./datasets/spotify_dataset.csv', )

<a id='2.1'></a>
## 2.1 Drop unused attributes (columns)
- Some of our attributes are useless, like: `Index`, `Song ID` and some are hard to use and have really little impact on our dataset (at least for my knowledge and experience at this point) like: `Week of Highest Charting`, `Genre`, `Weeks Charted` and `Chord`.

In [2]:
# Drop these columns all
df.drop(columns=['Index', 'Song ID', 'Week of Highest Charting', 'Genre', 'Weeks Charted', 'Chord'], inplace=True)

<a id='2.2'></a>
## 2.2 Repair Columns in wrong type
- Most attributes has the wrong data type, they should be number like `int` of `float`, not `object`: `Streams`, `Artist Followers`, `Popularity`, `Danceability`, `Energy`, `Louadness`, `Speechiness`, `Acousticness`, `Liveness`, `Tempo`, `Duration` and `Valence`. We have to change them.
- In columns that we want to repair, some of them ,`11` in particular, are strings with only `1` space (that's why when we call `null` function, it will return `0` null entry). We can't change these columns type without handle these 1-space strings, and we drop them for simplicity.

In [26]:
print('Number of Song with emty entries:', df[(df['Artist Followers'] == ' ')].shape[0])

Number of Song with emty entries: 11


In [31]:
# Drop theses emtyp columns (using there index)
df = df.drop(df[(df['Artist Followers'] == ' ')].index)

After drop these entries, we can now change columns' type.

In [40]:
# 'Streams' Attribute can't be turn into number since it has comma, remove them first
df['Streams'] = df['Streams'].replace(',', '', regex=True)

# These columns in int64 type
obj_to_int_cols = ["Streams", "Artist Followers", "Popularity"]

# And theses in float64
obj_to_float_cols = ["Danceability", "Energy", "Loudness", "Speechiness", "Acousticness",
                     "Liveness", "Tempo", "Duration (ms)", "Valence"]

df[obj_to_int_cols] = df[obj_to_int_cols].astype('int64')
df[obj_to_float_cols] = df[obj_to_float_cols].astype('float64')

In [41]:
# Check it one again
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1545 entries, 0 to 1555
Data columns (total 17 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Highest Charting Position  1545 non-null   int64  
 1   Number of Times Charted    1545 non-null   int64  
 2   Song Name                  1545 non-null   object 
 3   Streams                    1545 non-null   int64  
 4   Artist                     1545 non-null   object 
 5   Artist Followers           1545 non-null   int64  
 6   Release Date               1545 non-null   object 
 7   Popularity                 1545 non-null   int64  
 8   Danceability               1545 non-null   float64
 9   Energy                     1545 non-null   float64
 10  Loudness                   1545 non-null   float64
 11  Speechiness                1545 non-null   float64
 12  Acousticness               1545 non-null   float64
 13  Liveness                   1545 non-null   float

<a id='3'></a>
# 3 Data Overview
Take a quick look at our data to get a overal feel of our data.

**Observations:**
- Our data has `1545` rows represents `1545` songs, with `17` columns for each song's attributes.

<a id='3.1'></a>
## 3.1 Quick View
Below is the first 5 rows of our dataset:

In [3]:
df.head()

Highest Charting Position  Number of Times Charted  \
0                          1                        8   
1                          2                        3   
2                          1                       11   
3                          3                        5   
4                          5                        1   

                           Song Name     Streams          Artist  \
0                            Beggin'  48,633,449        Måneskin   
1          STAY (with Justin Bieber)  47,248,719   The Kid LAROI   
2                           good 4 u  40,162,559  Olivia Rodrigo   
3                         Bad Habits  37,799,456      Ed Sheeran   
4  INDUSTRY BABY (feat. Jack Harlow)  33,948,454       Lil Nas X   

  Artist Followers Release Date Popularity Danceability Energy Loudness  \
0          3377762   2017-12-08        100        0.714    0.8   -4.808   
1          2230022   2021-07-09         99        0.591  0.764   -5.484   
2          6266514   2021-05-21         99        0.563  0.664   -5.044   
3         83293380   2021-06-25         98        0.808  0.897   -3.712   
4          5473565   2021-07-23         96        0.736  0.704   -7.409   

  Speechiness Acousticness Liveness    Tempo Duration (ms) Valence  
0      0.0504        0.127    0.359  134.002        211560   0.589  
1      0.0483       0.0383    0.103  169.928        141806   0.478  
2       0.154        0.335   0.0849  166.928        178147   0.688  
3      0.0348       0.0469    0.364  126.026        231041   0.591  
4      0.0615       0.0203   0.0501  149.995        212000   0.894

And the last 5 one:

In [4]:
df.tail()

Highest Charting Position  Number of Times Charted  \
1551                        195                        1   
1552                        196                        1   
1553                        197                        1   
1554                        198                        1   
1555                        199                        1   

                               Song Name    Streams  \
1551                           New Rules  4,630,675   
1552                  Cheirosa - Ao Vivo  4,623,030   
1553           Havana (feat. Young Thug)  4,620,876   
1554          Surtada - Remix Brega Funk  4,607,385   
1555  Lover (Remix) [feat. Shawn Mendes]  4,595,450   

                             Artist Artist Followers Release Date Popularity  \
1551                       Dua Lipa         27167675   2017-06-02         79   
1552                 Jorge & Mateus         15019109   2019-10-11         66   
1553                 Camila Cabello         22698747   2018-01-12         81   
1554  Dadá Boladão, Tati Zaqui, OIK           208630   2019-09-25         60   
1555                   Taylor Swift         42227614   2019-11-13         70   

     Danceability Energy Loudness Speechiness Acousticness Liveness    Tempo  \
1551        0.762    0.7   -6.021      0.0694      0.00261    0.153  116.073   
1552        0.528   0.87   -3.123      0.0851         0.24    0.333   152.37   
1553        0.765  0.523   -4.333        0.03        0.184    0.132  104.988   
1554        0.832   0.55   -7.026      0.0587        0.249    0.182  154.064   
1555        0.448  0.603   -7.176       0.064        0.433   0.0862  205.272   

     Duration (ms) Valence  
1551        209320   0.608  
1552        181930   0.714  
1553        217307   0.394  
1554        152784   0.881  
1555        221307   0.422

In [42]:
print('Number of rows   : {}\nNumber of columns: {}'.format(df.shape[0], df.shape[1]))

Number of rows   : 1545
Number of columns: 17


<a id='3.2'></a>
## 3.2 Basic Statistics
Below is the basic statistics for each variables which contain information on `count`, `mean`, `standard deviation`, `minimum`, `1st quartile`, `median`, `3rd quartile`, and `maximum`.

In [38]:
df.describe()

Highest Charting Position  Number of Times Charted     Streams  \
count                    1545.00                  1545.00     1545.00   
mean                       87.83                    10.68  6337136.38   
std                        58.13                    16.39  3375402.18   
min                         1.00                     1.00  4176083.00   
25%                        37.00                     1.00  4915080.00   
50%                        80.00                     4.00  5269163.00   
75%                       137.00                    12.00  6452492.00   
max                       200.00                   142.00 48633449.00   

       Artist Followers  Popularity  Danceability  Energy  Loudness  \
count           1545.00     1545.00       1545.00 1545.00   1545.00   
mean        14716902.87       70.09          0.69    0.63     -6.35   
std         16675788.51       15.82          0.14    0.16      2.51   
min             4883.00        0.00          0.15    0.05    -25.17   
25%          2123734.00       65.00          0.60    0.53     -7.49   
50%          6852509.00       73.00          0.71    0.64     -5.99   
75%         22698747.00       80.00          0.80    0.75     -4.71   
max         83337783.00      100.00          0.98    0.97      1.51   

       Speechiness  Acousticness  Liveness   Tempo  Duration (ms)  Valence  
count      1545.00       1545.00   1545.00 1545.00        1545.00  1545.00  
mean          0.12          0.25      0.18  122.81      197940.82     0.51  
std           0.11          0.25      0.14   29.59       47148.93     0.23  
min           0.02          0.00      0.02   46.72       30133.00     0.03  
25%           0.05          0.05      0.10   97.96      169266.00     0.34  
50%           0.08          0.16      0.12  122.01      193591.00     0.51  
75%           0.17          0.39      0.22  143.86      218902.00     0.69  
max           0.88          0.99      0.96  205.27      588139.00     0.98